### PR TITLE
Fix missing parameter check in at::batch_norm

### LINKED ISSUE #117245: ```at::batch_norm```: Internal assert failed
### 🐛 Describe the bug

Test code:
```cpp
#include <torch/torch.h>

int main() {
  torch::TensorOptions toptions = torch::TensorOptions();
  toptions = toptions.dtype(torch::kComplexFloat);
  torch::Tensor input = torch::randn({1}, toptions);

  at::batch_norm(input, c10::nullopt, c10::nullopt, c10::nullopt, c10::nullopt, true, 1e-12, std::numeric_limits<double>::min(), false);

  return 0;
}
```

Error log:
```
[W SparseCsrTensorImpl.cpp:55] Warning: Sparse CSR tensor support is in beta state. If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues. (function operator())
terminate called after throwing an instance of 'c10::Error'
  what():  t->layout() == kStrided INTERNAL ASSERT FAILED at "/home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/MemoryOverlap.cpp":13, please report a bug to PyTorch. 
Exception raised from has_internal_overlap at /home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/MemoryOverlap.cpp:13 (most recent call first):
frame #0: <unknown function> + 0x1e7a6e (0x7f3779710a6e in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #1: <unknown function> + 0x1e7a20 (0x7f3779710a20 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #2: <unknown function> + 0x1e79b0 (0x7f37797109b0 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #3: <unknown function> + 0x1e7860 (0x7f3779710860 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #4: <unknown function> + 0x1e9286 (0x7f3779712286 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #5: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x1b7 (0x7f377970e4d7 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #6: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x259 (0x7f3779707ca9 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #7: ./reproduce_index_copy_() [0x40ee5e]
frame #8: at::has_internal_overlap(c10::TensorImpl*) + 0x189 (0x7f377c881d09 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #9: at::assert_no_internal_overlap(c10::TensorImpl*) + 0x15 (0x7f377c882165 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #10: at::assert_no_internal_overlap(at::TensorBase const&) + 0x1d (0x7f377c882acd in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #11: at::meta::structured_index_copy::meta(at::Tensor const&, long, at::Tensor const&, at::Tensor const&) + 0xc16 (0x7f377e66d1a6 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #12: <unknown function> + 0xab83904 (0x7f3784492904 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #13: <unknown function> + 0xab85dee (0x7f3784494dee in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x66d5397 (0x7f377ffe4397 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #15: <unknown function> + 0x66d4e82 (0x7f377ffe3e82 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #16: at::_ops::index_copy_::redispatch(c10::DispatchKeySet, at::Tensor&, long, at::Tensor const&, at::Tensor const&) + 0x543 (0x7f377fdc1623 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #17: <unknown function> + 0x153e4485 (0x7f378ecf3485 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #18: <unknown function> + 0x153e511b (0x7f378ecf411b in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #19: <unknown function> + 0x66d5397 (0x7f377ffe4397 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #20: <unknown function> + 0x66d4e82 (0x7f377ffe3e82 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #21: at::_ops::index_copy_::redispatch(c10::DispatchKeySet, at::Tensor&, long, at::Tensor const&, at::Tensor const&) + 0x543 (0x7f377fdc1623 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #22: <unknown function> + 0x1191f088 (0x7f378b22e088 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #23: <unknown function> + 0x11624b50 (0x7f378af33b50 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #24: <unknown function> + 0x11626e1b (0x7f378af35e1b in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #25: <unknown function> + 0x66d5397 (0x7f377ffe4397 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #26: at::_ops::index_copy_::call(at::Tensor&, long, at::Tensor const&, at::Tensor const&) + 0xe47 (0x7f377fa25c47 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #27: ./reproduce_index_copy_() [0x40972d]
frame #28: main + 0x13f8 (0x406938 in ./reproduce_index_copy_)
frame #29: <unknown function> + 0x29d90 (0x7f376fc4cd90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #30: __libc_start_main + 0x80 (0x7f376fc4ce40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #31: _start + 0x25 (0x405475 in ./reproduce_index_copy_)

Aborted (core dumped)
```

### Versions

PyTorch version: 2.3.0a0+git03969cb
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 12.0.1
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             96
On-line CPU(s) list:                0-95
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           7
CPU max MHz:                        4000.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6000.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           48 MiB (48 instances)
L3 cache:                           71.5 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-23,48-71
NUMA node1 CPU(s):                  24-47,72-95
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] optree==0.10.0
[pip3] torch==2.3.0a0+git03969cb
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.1.0         h06a4308_46344  
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] torch                     2.3.0a0+git03969cb           dev_0    <develop>

### PR DESCRIPTION
**Pull request by hipudding**
Input tensor valid checking are defined in
torch.nn.functional::batch_norm, if at::batch_norm is called, it may access illegal memory addresses.

Move valid checking to inner function to avoid memory corruption.

Fixes #117245

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
I don't remember where else the batch_norm() call can end up. We definitely want to make sure we cover all cases. @andrewor14 do you know off the top of your head if this move is enough or we need these checks somewhere else?

Also I guess these checks can also be removed from the python code that calls torch.batch_norm?

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -485,6 +485,23 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu_template(
 std::tuple<Tensor, Tensor, Tensor, Tensor, int64_t> _batch_norm_impl_index(
```
Comment (andrewor14) [inline aten/src/ATen/native/Normalization.cpp:485]: Moving forward, this won't be called after batch norm consolidation. I would recommend moving this check to the `batch_norm` function itself

**Comment by andrewor14** [comment]
> I don't remember where else the batch_norm() call can end up. We definitely want to make sure we cover all cases. @andrewor14 do you know off the top of your head if this move is enough or we need these checks somewhere else?
> 
> Also I guess these checks can also be removed from the python code that calls torch.batch_norm?

These checks should be enough for the current state, since both eager and compile/export go through this op. This won't be true after batch norm consolidation, however. A better place to do these checks would be in `aten.batch_norm` itself; that will always be the common aten op for both flows

**Comment by andrewor14** [comment]
> I don't remember where else the batch_norm() call can end up. We definitely want to make sure we cover all cases. @andrewor14 do you know off the top of your head if this move is enough or we need these checks somewhere else?
> 
> Also I guess these checks can also be removed from the python code that calls torch.batch_norm?

These checks should be enough for the current state, since both eager and compile/export go through this op. However, this will no longer be true after batch norm consolidation. A better place to do these checks would be in `aten.batch_norm` itself; that will always be the common aten op for both flows

**Comment by hipudding** [comment]
Thanks @albanD @andrewor14 for your review. I was exciting that this is my first contribute to pytorch.

I followed call stack of batch_norm, there's a lot dispatcher stack, and I thought _batch_norm_impl_index is the first not generate function or dispatch function. I found aten.batch_norm function are generated by gen.sh, do you mean I should to change something in code generation?

**Comment by albanD** [comment]
From the CI signal, it does seem like these tests are too restrictive though :/

**Comment by andrewor14** [comment]
> Thanks @albanD @andrewor14 for your review. I was exciting that this is my first contribute to pytorch.
> 
> I followed call stack of batch_norm, there's a lot dispatcher stack, and I thought _batch_norm_impl_index is the first not generate function or dispatch function. I found aten.batch_norm function are generated by gen.sh, do you mean I should to change something in code generation or something in code template?

I don't think `aten.batch_norm` is generated? Can you add the check here? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Normalization.cpp#L640

**Comment by hipudding** [comment]
> From the CI signal, it does seem like these tests are too restrictive though :/ 
> 
> 

Thanks. I will check it.

**Comment by hipudding** [comment]
> > Thanks @albanD @andrewor14 for your review. I was exciting that this is my first contribute to pytorch.
> 
> > 
> 
> > I followed call stack of batch_norm, there's a lot dispatcher stack, and I thought _batch_norm_impl_index is the first not generate function or dispatch function. I found aten.batch_norm function are generated by gen.sh, do you mean I should to change something in code generation or something in code template?
> 
> 
> 
> I don't think `aten.batch_norm` is generated? Can you add the check here? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Normalization.cpp#L640

Got it. Thanks.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -413,8 +413,6 @@ def sample_inputs_batch_norm(op_info, device, dtype, requires_grad, **kwargs):
         ((S, S, S), {'training': True, 'momentum': 0.5, 'eps': 0.6}),
         ((3, 2, 4), {'training': False, 'momentum': -1.2}),
         ((3, 1), {'training': True, 'momentum': 0.0}),
-        ((0,), {'training': True}),
-        ((0,), {'training': False}),
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:417]: Why are you removing these test cases?

Comment (hipudding) [reply torch/testing/_internal/common_methods_invocations.py:417]: According to the content of this PR modification, batch_norm cannot accept tensors less than 2dims, so these two input data will cause many test cases failed.
I have already add a testcase to cover tensor less than 2dims.

Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:417]: The thing is that these two tests used to pass. So if we're saying they're not passing anymore, this would be BC-breaking.
I don't think we should change the behavior of batchnorm in any way here.

We would need to make sure the test only prevent the case that segfault but still allow anything that used to be valid before.

Comment (hipudding) [reply torch/testing/_internal/common_methods_invocations.py:417]: Thank you for your reply. But the fact is the only thing I have done in this PR is move input parameters check from outer function to barch_norm itself, because some times, this parameter check is not called(the check is not in every call stack).  I the past, these two testcase can pass because batch_norm can handle empty tensor(it will return another empty tensor as the result), but even so, some code has already visit invalid memory(such as here: https://github.com/pytorch/pytorch/blob/677e67c3998df11bee0d371f9f6f9482c9e4c29e/aten/src/ATen/native/Normalization.cpp#L495, if input only as 1 dim, input.sym_sizes()[1] is not a valid address). That why I create this PR to prevent input less than 2 dims.

For these two test cases, input only has 1 dim, it will checked by new added parameter check, So I remove them. 

Do you have any good suggestions to keep these two test cases but fix this issue? Thanks.

Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:417]: Someone would need to dive into details of what is expected or not for the python side of things but the sample just above this one is `((3, 1), {'training': True, 'momentum': 0.0}),` which would mean a 2D input IIRC.

I'm not saying this is bug free but the key expectation for this code is stability. So I do not want to make any change that might change what works or not without understanding in details the state of what is expected.
Sorry for dragging this PR but this is a very high profile function and we definitely don't want to break anything that is working today.

Comment (hipudding) [reply torch/testing/_internal/common_methods_invocations.py:417]: Thanks. I got it. I will try to not break anything and fix this issue.
